src/main/java/net/dv8tion/jda/api/utils/ConcurrentSessionController.java

```diff
@@ -0,0 +1,159 @@
+    {
+        // assertions are ignored at runtime by default, this is a sanity check
+        assert level > 0 && level < Integer.MAX_VALUE;
+        workers = new Worker[level];
+    }
+
+    @Override
+    public void appendSession(@Nonnull SessionConnectNode node)
+    {
+        getWorker(node).enqueue(node);
+    }
+
+    @Override
+    public void removeSession(@Nonnull SessionConnectNode node)
+    {
+        getWorker(node).dequeue(node);
+    }
+
+    private synchronized Worker getWorker(SessionConnectNode node)
+    {
+        // get or create worker (synchronously since this should be thread-safe)
+        int i = node.getShardInfo().getShardId() % workers.length;
+        Worker worker = workers[i];
+        if (worker == null)
+        {
+            log.debug("Creating new worker handle for shard pool {}", i);
+            workers[i] = worker = new Worker(i);
+        }
+        return worker;
+    }
+
+    private static class Worker extends Thread
+    {
+        private final Queue<SessionConnectNode> queue = new ConcurrentLinkedQueue<>();
+
+        public Worker(int id)
+        {
+            super("ConcurrentSessionController-Worker-" + id);
+        }
+
+        public void enqueue(SessionConnectNode node)
+        {
+            log.trace("Appending node to queue {}", node.getShardInfo());
+            queue.add(node);
+            execute();
+        }
+
+        public void dequeue(SessionConnectNode node)
+        {
+            log.trace("Removing node from queue {}", node.getShardInfo());
+            queue.remove(node);
+        }
+
+        public void execute()
+        {
+            if (!isAlive())
+            {
+                log.debug("Running worker");
+                start();
+            }
+        }
+
+        @Override
+        public void run()
+        {
+            try
+            {
+                while (!queue.isEmpty())
+                {
+                    processQueue();
+                    // We always sleep here because its possible that we get a new session request before the rate limit expires
+                    TimeUnit.SECONDS.sleep(SessionController.IDENTIFY_DELAY);
+                }
+            }
+            catch (InterruptedException ex)
+            {
+                log.error("Worker failed to process queue", ex);
+            }
+        }
+
+        private void processQueue() throws InterruptedException
+        {
+            SessionConnectNode node = null;
+            try
+            {
+                node = queue.remove();
+                log.debug("Running connect node for shard {}", node.getShardInfo());
+                node.run(false); // we don't use isLast anymore because it can be a problem with many reconnecting shards
+            }
+            catch (NoSuchElementException ignored) {/* This means the node was removed before we started it */}
+            catch (InterruptedException e)
+            {
+                queue.add(node);
```

Comment:
We are readding the node to the queue but still throwing the exception. This exception will kill the thread.
Are we just relying on the next call to `enqueue` to restart the thread?

Reply:
Yes. The thread can only be interrupted by a third-party so it must be either intentional or a bug. Ideally, the third-party will also handle starting a new one in that case.

Reply:
Okay, just wanted to make sure we were on the same page. 👍